Add federation support

kindermax opened the pull request:
This is a draft PR which adds apollo federation support to hiku.

Union type is required to implement federation, so I've added it into hiku internals.

Federation can be implemented either as a separate package or inside hiku.

What is added:
- Added Union type
- Added federation support https://www.apollographql.com/docs/federation/federation-spec/ 
- Added example for federation (example server works as a server which exposes Astronaut entity with "id" as its primary key)
- Added several tests for federation (will add more if we will agree on an idea and implementation)

What is not ready:
 - https://www.apollographql.com/docs/federation/federation-spec/#requires
 - https://www.apollographql.com/docs/federation/federation-spec/#provides
 - https://www.apollographql.com/docs/federation/federation-spec/#external
 
There are some questions about the proper approach for implementing Unions.
And also this code may be written by not following hiku idioms, but I am ready to work for a better solution.
Thanks

kindermax commented on federation/entities.py:
```diff
@@ -0,0 +1,40 @@
+    _entity_fields = entities_link.node.fields
+    representations = entities_link.options.get('representations')
+
+    entity_type = representations[0]['__typename']
```
Apollo Gateway splits representations by __typename, so it is kind of safe to handle just one resolver.
But theoretically, there can be several distinct __typename in representations (I've done it by querying _entities directly on the service graph).

Some implementations in kotlin and graphene for example:
- https://github.com/ExpediaGroup/graphql-kotlin/blob/master/graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/execution/EntityResolver.kt
- https://github.com/preply/graphene-federation/blob/master/graphene_federation/entity.py

kindermax commented on federation/graph.py:
```diff
@@ -0,0 +1,108 @@
+        )
+
+    def get_root_entities_node(self, items):
+        return Field(
```
I had an idea to use a Link here, but it requires many internal changes to be made inside hiku to link to the nodes from Union dynamically

vmagamedov commented:
`Union` type can't be added right now, as you said, this requires a lot of changes. Mostly because of how link resolvers are designed.

I think that federation can be added on top of hiku, with original resolvers. So there is no need to use `FederatedResolver` class.

Federation is mostly about SDL, directives and one special `_entities` resolver. I think that you can extend or rewrite hiku's GraphQL introspection implementation to support SDL and directives. And override `Engine` class to execute GraphQL query a little bit differently.

Generally hiku executes query from the root node. Here is how you can change this strategy and process specific nodes with specific ids (this is what `_entities` resolver should do). Replace
```python
query_workflow.start()
```
with something like this
```python
for node, node_query, node_ids in representations:
    query_workflow.process_node(node, node_query, node_ids)
```
where
```graphql
query ($_representations: [_Any!]!) {
     # ^^^^^^^^^^^^^^^^^ node_ids
  _entities(representations: $_representations) {
    ... on Product {
      reviews {  # <-- node_query
        body
      }
    }
  }
}
```
After executing this query, you can access all field values in the result object. Then you will need to properly denormalize result into what federation spec expects.

kindermax commented:
Thanks for the response. I will try to rewrite some parts of PR and will see where it goes.

As for
> I think that you can extend or rewrite hiku's GraphQL introspection implementation to support SDL and directives.
If I understand you correctly, this will allow me to add unions.. without adding actual unions ? In other words, is there are chances to implement federation without actual support of Unions ?

vmagamedov commented:
Our introspection implementation wasn't written with extensibility in mind, but I'm ok to make it extensible if possible.

I'm also ok to add federation support into hiku, but I want it to be implemented on top of core functionality, with loose coupling.

Yes, I suggest you to use `Union` without actually adding the `Union` type into hiku :) Federation-aware introspection will extend original graph introspection with this:
```graphql
scalar _Any
scalar _FieldSet

union _Entity = User | Product  # <-- generated from graph nodes

type _Service {
  sdl: String
}

extend type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

directive @external on FIELD_DEFINITION
directive @requires(fields: _FieldSet!) on FIELD_DEFINITION
directive @provides(fields: _FieldSet!) on FIELD_DEFINITION
directive @key(fields: _FieldSet!) on OBJECT | INTERFACE

# this is an optional directive discussed below
directive @extends on OBJECT | INTERFACE
```

`_entities` field shouldn't be added into the graph definition, federation-aware engine should process this field by itself (as in my previous comment), because this field is very special and can't be processed as other regular fields in the graph. So in this way we don't have to add unions into our graph definition.

kindermax commented:
@vmagamedov Hi. I did a complete rewrite of the federation implementation. It would be great if will take a look :)

- Introspection will be extended with federation's types in a bit hacky way but this allows to keep core almost unchanged (and no unions needed)
- Refactored directives in introspection, added directives to Node and Field. This is the most important change, as directives give us a lot of flexibility in how to work with graph and its items.
- New federation-aware engine, handles `_service` and `_entities` queries differently
- New federation-aware endpoints with custom denormalization logic
- Sdl printer based on graphql ast instead of string concatenation

vmagamedov commented:
@kindermax Hi, impressive work, sure, I'll take a look

kindermax commented on hiku/federation/endpoint.py:
```diff
@@ -0,0 +1,155 @@
+                continue
+            ident = r[key]
+            result.__ref__ = Reference(typename, ident)
+            data = DenormalizeGraphQL(graph, result, type_name).process(node)
```
Whoops. This is broken (works by coincedence). Turns out, I had a link `cart` in Root node, and a query `cart { id }`, so `self._type[-1].__field_types__[obj.name]` worked fine. This should be fixed somehow.

kindermax replied:
fixed by using type from representations as root type for denormalization instead of `__root__`

vmagamedov commented:
Would be nice to avoid having `directives` argument in the classes in the `hiku.graph` module. I'm currently thinking about some sort of `metadata` argument for extensibility in general, but don't see yet how to do this better. Want to have extensibility and ease of use at the same time.

kindermax commented on hiku/graph.py:
```diff
@@ -131,6 +133,10 @@ def __repr__(self):
         return OrderedDict((op.name, op) for op in self.options)
 
+    @cached_property
+    def directives_map(self):
```
According to federation spec, https://www.apollographql.com/docs/federation/federation-spec/#key, there can be multiple same directives, so directives_map is definitely a wrong decision.

vmagamedov replied:
And `_FieldSet` type can have multiple values, even nested 🤯 I wonder why they don't use `String` type here with ability to specify several directives

kindermax replied:
I haven't implemented multiple and nested values. For now just single field like `id` or `cartId`

kindermax replied:
But I assume, since _FieldSet is a selection set, we have to parse fields string into selection set

vmagamedov replied:
> I haven't implemented multiple and nested values. For now just single field like `id` or `cartId`

Probably it would be enough to start with.

> But I assume, since _FieldSet is a selection set, we have to parse fields string into selection set

I was thinking that `@key` directive is more for the federation gateway service, not for us. Maybe I'm missing something, I'm not finished reviewing this PR.

kindermax replied:
@key directive is a way of saying which field in type has to be a primary key. In prom, we have Order and Cart types, and since these types are in different services, we have to use @key to define primary keys.

kindermax replied:
here is an example how we can extend some types from one service in another. Here we are extending `Order` node with `cart` field, and declaring `Product` stub type, so we can add `product` field into `CartItem` type in cart service
```

extend_nodes = [
    Node('Order', [
        Field('cartId', Integer, ids_resolver,
              directives=[External()]),
        Link('cart', TypeRef['Cart'], direct_link, requires='cartId'),
    ], directives=[Key('cartId'), Extends()]),
    Node('Product', [
        Field('id', Integer, ids_resolver, directives=[External()]),
    ], directives=[Key('id'), Extends()]),
]

```

kindermax commented:
> Would be nice to avoid having directives argument in the classes in the hiku.graph module
For not I have no any good ideas how to do it differently

kindermax commented:
Also noticed that for example `@key` directive is repeatable (can be used several times for one type), but `@deprecated` directive is not repeatable. So, we have to figure out how to validate this during graph creation

kindermax commented:
@vmagamedov Hi, i'm curious if there is something we have to do before merging this PR or its ready ?

vmagamedov commented:
@kindermax I think that this PR is good enough, but maybe it can be polished in the future. My main concern was the public API exposed to the users. It is harder to change API than to change/fix implementation details in the future.

I also didn't had time to completely review this PR, only partly. And I think that I'm significantly slowing down your progress. So, since this project is mainly used by one company, I think that it would be right to:
1. give you access rights to this repository to commit/merge/release new versions
2. transfer this project to the `evo-company` org
3. transfer ownership for the `hiku` package on the PyPI

kindermax commented:
I think that we can merge this PR and release some draft version, use this version for some time, polish public API and then after some iterations, release new version.

As for the ownership question, for now, access rights to the repository is enough for now to proceed further with the federation.
As to `evo-company`, I think it is a good move but I can not solely make that decision (but I will talk to `evo-company` org members), so we can make a decision later in another issue.

vmagamedov commented:
I've already sent you an invitation. Try to merge this PR to check your access rights. Then you need to update version here https://github.com/vmagamedov/hiku/blob/edcd3c177044010ecd11b1bf33fa43a40531605c/hiku/__init__.py#L1 and tag this commit. Release workflow should work fine.

kindermax commented:
I merged PR successfully. Thank you for you support in implementing this feature.
